src/dist/replication/test/replica_test/unit_test/replica_disk_test.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017-present, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include <gtest/gtest.h>
+
+#include <dsn/utility/fail_point.h>
+#include "replica_test_base.h"
+
+namespace dsn {
+namespace replication {
+
+class replica_disk_test : public testing::Test
+{
+public:
+    std::unique_ptr<mock_replica_stub> _stub;
+    int app_id = 1;
+    int dir_nodes_count = 2;
+    int primary_count_for_disk = 1;
+    int secondary_count_for_disk = 2;
+    dsn::app_info app_info;
+
+public:
+    void SetUp()
+    {
+        _stub = make_unique<mock_replica_stub>();
+        generate_mock_app_info();
+        generate_mock_dir_nodes(dir_nodes_count);
+        _stub->generate_replicas_base_dir_nodes_for_app(
+            app_info, primary_count_for_disk, secondary_count_for_disk);
+        _stub->on_disk_stat();
+    }
+
+    std::vector<std::shared_ptr<dir_node>> get_fs_manager_nodes()
+    {
+        return _stub->_fs_manager._dir_nodes;
+    }
+
+    void TearDown() {}
+
+private:
+    void generate_mock_app_info()
+    {
+        app_info.app_id = app_id;
+        app_info.app_name = "disk_test";
+        app_info.app_type = "replica";
+        app_info.is_stateful = true;
+        app_info.max_replica_count = 3;
+        app_info.partition_count = 8;
+    }
+
+    void generate_mock_dir_nodes(int num)
+    {
+        int64_t disk_capacity_mb = num * 100;
+        while (num > 0) {
+            int64_t disk_available_mb = num * 10;
+            int64_t disk_available_ratio = static_cast<int64_t>(
+                std::round((double)100 * disk_available_mb / disk_capacity_mb));
+            // create one mock dir_node and make sure disk_capacity_mb_ > disk_available_mb_
+            dir_node *node_disk = new dir_node("tag_" + std::to_string(num),
+                                               "full_dir_" + std::to_string(num),
+                                               disk_capacity_mb,
+                                               disk_available_mb,
+                                               disk_available_ratio);
+            int replica_index = 0;
+            int disk_holding_replica_count = primary_count_for_disk + secondary_count_for_disk;
+            while (disk_holding_replica_count-- > 0) {
+                node_disk->holding_replicas[app_id].emplace(gpid(app_id, replica_index++));
+            }
+            _stub->_fs_manager._dir_nodes.emplace_back(node_disk);
```

Comment:
`node_disk`s are put into `_dir_nodes` regularly, i.e. in inverted order of `disk_available_mb`, could you shuffle the order?

Reply:
Only `_stub->on_disk_stat()` can trigger shuffle, In there, `_stub->on_disk_stat()`  is excuted in `SetUp()` in `replica_disk_test.cpp`